File: factorial/factorial.c
#include <stdlib.h>
#include <stdio.h>

long factorial(int x) {

  int i;              /* counter */
  long this_fact;     /* factorial for this counter */
  long last_fact = 1; /* factorial for last counter */

  if (x < 2) {
    return 1;
  }
  
  for (i = 2; i <= x; i++) {
    this_fact = i * last_fact;
    last_fact = this_fact;
  }

  return this_fact;
}

/*
 * This doesn't exactly work because strtol cannot differentiate between
 * something that doesn't convert to a number and an actual zero; strtol
 * returns 0 for both.
 */
int main(int argc, char *argv[]) {
  int x;
  long fact;

  if (argc < 2) {
    printf("Usage: %s <non-negative number>\n", argv[0]);
  } else {
    x = strtol(argv[1], NULL, 10);
    if (x <= 0) {
      printf("Usage: %s <non-negative number>\n", argv[0]);
    } else {
      fact = factorial(x);
      printf("The factorial of %d is: %ld\n", x, fact);
    }
  }
}

File: factorial/factorial.rb
#!/usr/bin/env ruby 

def factorial(x)
  return "<Error: input is negative>" unless x >= 0
  return 1 unless x >= 2

  this_fact = nil
  last_fact = 1
  (2..x).each do |i|
    this_fact = i * last_fact
    last_fact = this_fact
  end

  return this_fact
end

unless ARGV[0] then
  puts "Usage: factorial.rb <non-negative integer>"
else
  puts "The factorial of " + ARGV[0] + " is: " + factorial(ARGV[0].to_i).to_s
end